trl/trainer/sft_trainer.py

```diff
@@ -0,0 +1,310 @@
+            if dataset_text_field is None and formatting_func is None:
+                raise ValueError(
+                    "You passed `packing=False` to the SFTTrainer, but you didn't pass a `dataset_text_field` or `formatting_func` argument."
+                )
+
+            if data_collator is None:
+                data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=False)
+
+        if train_dataset is not None:
+            train_dataset = self._prepare_dataset(
+                train_dataset,
+                tokenizer,
+                packing,
+                dataset_text_field,
+                max_seq_length,
+                formatting_func,
+                infinite,
+                num_of_sequences,
+                chars_per_token,
+            )
+        if eval_dataset is not None:
+            eval_dataset = self._prepare_dataset(
+                eval_dataset,
+                tokenizer,
+                packing,
+                dataset_text_field,
+                max_seq_length,
+                formatting_func,
+                infinite,
+                num_of_sequences,
+                chars_per_token,
+            )
+
+        super().__init__(
+            model=model,
+            args=args,
+            data_collator=data_collator,
+            train_dataset=train_dataset,
+            eval_dataset=eval_dataset,
+            tokenizer=tokenizer,
+            model_init=model_init,
+            compute_metrics=compute_metrics,
+            callbacks=callbacks,
+            optimizers=optimizers,
+            preprocess_logits_for_metrics=preprocess_logits_for_metrics,
+        )
+
+    def _prepare_dataset(
+        self,
+        dataset,
+        tokenizer,
+        packing,
+        dataset_text_field,
+        max_seq_length,
+        formatting_func,
+        infinite,
+        num_of_sequences,
+        chars_per_token,
+    ):
+        # check if torch dataset / dataloader and do nothing
+        if dataset is not None and (
+            isinstance(
+                dataset,
+                (torch.utils.data.IterableDataset, torch.utils.data.Dataset),
+            )
+        ):
+            is_already_dataset = True
+        else:
+            is_already_dataset = False
+
+        if not packing:
+            dataset = self._prepare_non_packed_dataloader(
+                tokenizer, dataset, dataset_text_field, max_seq_length, formatting_func
+            )
+
+            is_already_dataset = True
+
+        if not is_already_dataset and (dataset_text_field is not None or formatting_func is not None):
+            if tokenizer is None:
+                raise ValueError(
+                    "You need to pass a tokenizer when using the SFT Trainer when passing a `dataset_text_field`."
+                )
+
+            dataset = ConstantLengthDataset(
+                tokenizer,
+                dataset[dataset_text_field] if dataset_text_field is not None else dataset,
```

Comment:
this will load the full dataset into memory. i think we should rather pass the column name to the `ConstantLengthDataset`. In the CLD: if a column name is provided then `self.formatting_func = lambda x: x[column_name]` and else the prerpocessing function. What do you think?

Reply:
Sounds great!